### ISSUE TITLE
Errno::EBADF Bad file descriptor - not a socket file descriptor

### ISSUE THREAD

**Issue by ryan-ol** [puma/puma#1046]
Puma 3.6.0 / Ruby 2.3.0-p0/ Rails 5.0.0.1. When running 'rails restart' I encounter the issue below.

I checked out [https://github.com/puma/puma-dev/issues/12](https://github.com/puma/puma-dev/issues/12) and was unsuccessful in solving the issue. 

I also attempted using the master github branch but was still unable to restart using the new shortcut. Here is what I see:

```
Exiting
/Users/name/.rvm/gems/ruby-2.3.0/bundler/gems/puma-ddea4f345d17/lib/puma/binder.rb:284:in `for_fd': Bad file descriptor - not a socket file descriptor (Errno::EBADF)
	from /Users/name/.rvm/gems/ruby-2.3.0/bundler/gems/puma-ddea4f345d17/lib/puma/binder.rb:284:in `inherit_tcp_listener'
	from /Users/name/.rvm/gems/ruby-2.3.0/bundler/gems/puma-ddea4f345d17/lib/puma/binder.rb:91:in `block in parse'
	from /Users/name/.rvm/gems/ruby-2.3.0/bundler/gems/puma-ddea4f345d17/lib/puma/binder.rb:85:in `each'
	from /Users/name/.rvm/gems/ruby-2.3.0/bundler/gems/puma-ddea4f345d17/lib/puma/binder.rb:85:in `parse'
	from /Users/name/.rvm/gems/ruby-2.3.0/bundler/gems/puma-ddea4f345d17/lib/puma/runner.rb:133:in `load_and_bind'
	from /Users/name/.rvm/gems/ruby-2.3.0/bundler/gems/puma-ddea4f345d17/lib/puma/single.rb:85:in `run'
	from /Users/name/.rvm/gems/ruby-2.3.0/bundler/gems/puma-ddea4f345d17/lib/puma/launcher.rb:172:in `run'
	from /Users/name/.rvm/gems/ruby-2.3.0/bundler/gems/puma-ddea4f345d17/lib/rack/handler/puma.rb:51:in `run'
	from /Users/name/.rvm/gems/ruby-2.3.0/gems/rack-2.0.1/lib/rack/server.rb:296:in `start'
	from /Users/name/.rvm/gems/ruby-2.3.0/gems/railties-5.0.0.1/lib/rails/commands/server.rb:79:in `start'
	from /Users/name/.rvm/gems/ruby-2.3.0/gems/railties-5.0.0.1/lib/rails/commands/commands_tasks.rb:90:in `block in server'
	from /Users/name/.rvm/gems/ruby-2.3.0/gems/railties-5.0.0.1/lib/rails/commands/commands_tasks.rb:85:in `tap'
	from /Users/name/.rvm/gems/ruby-2.3.0/gems/railties-5.0.0.1/lib/rails/commands/commands_tasks.rb:85:in `server'
	from /Users/name/.rvm/gems/ruby-2.3.0/gems/railties-5.0.0.1/lib/rails/commands/commands_tasks.rb:49:in `run_command!'
	from /Users/name/.rvm/gems/ruby-2.3.0/gems/railties-5.0.0.1/lib/rails/commands.rb:18:in `<top (required)>'
	from /Users/name/Desktop/airbnb/bin/rails:9:in `require'
	from /Users/name/Desktop/airbnb/bin/rails:9:in `<top (required)>'
	from /Users/name/.rvm/gems/ruby-2.3.0/gems/spring-1.7.2/lib/spring/client/rails.rb:28:in `load'
	from /Users/name/.rvm/gems/ruby-2.3.0/gems/spring-1.7.2/lib/spring/client/rails.rb:28:in `call'
	from /Users/name/.rvm/gems/ruby-2.3.0/gems/spring-1.7.2/lib/spring/client/command.rb:7:in `call'
	from /Users/name/.rvm/gems/ruby-2.3.0/gems/spring-1.7.2/lib/spring/client.rb:30:in `run'
	from /Users/name/.rvm/gems/ruby-2.3.0/gems/spring-1.7.2/bin/spring:49:in `<top (required)>'
	from /Users/name/.rvm/gems/ruby-2.3.0/gems/spring-1.7.2/lib/spring/binstub.rb:11:in `load'
	from /Users/name/.rvm/gems/ruby-2.3.0/gems/spring-1.7.2/lib/spring/binstub.rb:11:in `<top (required)>'
	from /Users/name/Desktop/airbnb/bin/spring:13:in `require'
	from /Users/name/Desktop/airbnb/bin/spring:13:in `<top (required)>'
	from bin/rails:3:in `load'
	from bin/rails:3:in `<main>'
```

**Comment by evanphx** [comment]
Did you update to the latest version of puma-dev? That was the fix for that issue.

**Comment by ryan-ol** [comment]
@evanphx yes I installed the latest version of puma-dev and received the error above

**Comment by evanphx** [comment]
Can you run 'puma-dev -V' and also check the full path to puma-dev running
via 'ps'

**Comment by ryan-ol** [comment]
@evanphx 

```
MacBook-Pro:app-name name$ puma-dev -V
Version: v0.10 (go1.7)
MacBook-Pro:app-name name$ ps
  PID TTY           TIME CMD
54326 ttys000    0:00.12 -bash
69600 ttys000    0:09.51 puma 3.6.0 (tcp://localhost:3000) [app-name]  
69603 ttys000    0:00.01 /Users/name/.rvm/gems/ruby-2.3.0/gems/rb-fsevent-0.9.7/bin/fsevent_watch --latency 0.1 /Users/name/.rvm/gems/ruby-2.3
69604 ttys000    0:00.01 /Users/name/.rvm/gems/ruby-2.3.0/gems/rb-fsevent-0.9.7/bin/fsevent_watch --latency 0.1 /Users/name/.rvm/gems/ruby-2.3
69605 ttys000    0:00.01 /Users/name/.rvm/gems/ruby-2.3.0/gems/rb-fsevent-0.9.7/bin/fsevent_watch --latency 0.1 /Users/name/.rvm/gems/ruby-2.3
69606 ttys000    0:00.01 /Users/name/.rvm/gems/ruby-2.3.0/gems/rb-fsevent-0.9.7/bin/fsevent_watch --latency 0.1 /Users/name/.rvm/gems/ruby-2.3
69607 ttys000    0:00.01 /Users/name/.rvm/gems/ruby-2.3.0/gems/rb-fsevent-0.9.7/bin/fsevent_watch --latency 0.1 /Users/name/.rvm/gems/ruby-2.3
69608 ttys000    0:00.01 /Users/name/.rvm/gems/ruby-2.3.0/gems/rb-fsevent-0.9.7/bin/fsevent_watch --latency 0.1 /Users/name/.rvm/gems/ruby-2.3
69609 ttys000    0:00.01 /Users/name/.rvm/gems/ruby-2.3.0/gems/rb-fsevent-0.9.7/bin/fsevent_watch --latency 0.1 /Users/name/Desktop/app-name/con
69610 ttys000    0:00.01 /Users/name/.rvm/gems/ruby-2.3.0/gems/rb-fsevent-0.9.7/bin/fsevent_watch --latency 0.1 /Users/name/.rvm/gems/ruby-2.3
69611 ttys000    0:00.01 /Users/name/.rvm/gems/ruby-2.3.0/gems/rb-fsevent-0.9.7/bin/fsevent_watch --latency 0.1 /Users/name/.rvm/gems/ruby-2.3
69612 ttys000    0:00.01 /Users/name/.rvm/gems/ruby-2.3.0/gems/rb-fsevent-0.9.7/bin/fsevent_watch --latency 0.1 /Users/name/.rvm/gems/ruby-2.3
69613 ttys000    0:00.01 /Users/name/.rvm/gems/ruby-2.3.0/gems/rb-fsevent-0.9.7/bin/fsevent_watch --latency 0.1 /Users/name/.rvm/gems/ruby-2.3
69614 ttys000    0:00.01 /Users/name/.rvm/gems/ruby-2.3.0/gems/rb-fsevent-0.9.7/bin/fsevent_watch --latency 0.1 /Users/name/.rvm/gems/ruby-2.3
69615 ttys000    0:00.01 /Users/name/.rvm/gems/ruby-2.3.0/gems/rb-fsevent-0.9.7/bin/fsevent_watch --latency 0.1 /Users/name/.rvm/gems/ruby-2.3
69616 ttys000    0:00.01 /Users/name/.rvm/gems/ruby-2.3.0/gems/rb-fsevent-0.9.7/bin/fsevent_watch --latency 0.1 /Users/name/.rvm/gems/ruby-2.3
69618 ttys000    0:00.01 /Users/name/.rvm/gems/ruby-2.3.0/gems/rb-fsevent-0.9.7/bin/fsevent_watch --latency 0.1 /Users/name/Desktop/app-name/db
69619 ttys000    0:00.01 /Users/name/.rvm/gems/ruby-2.3.0/gems/rb-fsevent-0.9.7/bin/fsevent_watch --latency 0.1 /Users/name/Desktop/app-name/app
69620 ttys000    0:00.01 /Users/name/.rvm/gems/ruby-2.3.0/gems/rb-fsevent-0.9.7/bin/fsevent_watch --latency 0.1 /Users/name/Desktop/app-name/app
69621 ttys000    0:00.01 /Users/name/.rvm/gems/ruby-2.3.0/gems/rb-fsevent-0.9.7/bin/fsevent_watch --latency 0.1 /Users/name/Desktop/app-name/app
69622 ttys000    0:00.01 /Users/name/.rvm/gems/ruby-2.3.0/gems/rb-fsevent-0.9.7/bin/fsevent_watch --latency 0.1 /Users/name/Desktop/app-name/app
69623 ttys000    0:00.01 /Users/name/.rvm/gems/ruby-2.3.0/gems/rb-fsevent-0.9.7/bin/fsevent_watch --latency 0.1 /Users/name/Desktop/app-name/app
69624 ttys000    0:00.01 /Users/name/.rvm/gems/ruby-2.3.0/gems/rb-fsevent-0.9.7/bin/fsevent_watch --latency 0.1 /Users/name/Desktop/app-name/app
69625 ttys000    0:00.01 /Users/name/.rvm/gems/ruby-2.3.0/gems/rb-fsevent-0.9.7/bin/fsevent_watch --latency 0.1 /Users/name/Desktop/app-name/app
69626 ttys000    0:00.01 /Users/name/.rvm/gems/ruby-2.3.0/gems/rb-fsevent-0.9.7/bin/fsevent_watch --latency 0.1 /Users/name/.rvm/gems/ruby-2.3
69627 ttys000    0:00.01 /Users/name/.rvm/gems/ruby-2.3.0/gems/rb-fsevent-0.9.7/bin/fsevent_watch --latency 0.1 /Users/name/.rvm/gems/ruby-2.3
69628 ttys000    0:00.01 /Users/name/.rvm/gems/ruby-2.3.0/gems/rb-fsevent-0.9.7/bin/fsevent_watch --latency 0.1 /Users/name/.rvm/gems/ruby-2.3
69629 ttys000    0:00.01 /Users/name/.rvm/gems/ruby-2.3.0/gems/rb-fsevent-0.9.7/bin/fsevent_watch --latency 0.1 /Users/name/Desktop/app-name/tes
55531 ttys001    0:00.15 -bash
MacBook-Pro:app-name name$ 
```

**Comment by evanphx** [comment]
@ryan-ol Could you run `ps ax | grep puma` for me? Puma-dev isn't showing up at all in the normal `ps` because of the way process sessions work.

**Comment by ryan-ol** [comment]
@evanphx sure:

```
MacBook-Pro:User user$ ps ax | grep puma
46155   ??  S      0:06.45 /usr/local/Cellar/puma-dev/0.10/bin/puma-dev -launchd -dir ~/.puma-dev -d dev -timeout 15m0s
69600 s000  S+     0:16.69 puma 3.6.0 (tcp://localhost:3000) [appname]  
99391 s003  S+     0:00.00 grep puma
MacBook-Pro:User user$ 
```

**Comment by evanphx** [comment]
You have something very odd going on with spring because puma-dev isn't using that mechanism anymore, no should the app be launched via `rails s` which it appears to be in the backtrace.

Oh, I reread your docs and you're using `rails restart`. I'm not certain that will work properly, try using the native mechanism of touching `tmp/restart.txt`.

**Comment by ericboehs** [comment]
@evanphx I have encountered this problem as well.

Rails 5.0.0.1
puma 3.6.0
No puma-dev
macOS Sierra

Here's my `ps ax | grep puma`:
```
 3335 s002  S+     0:01.89 puma 3.6.0 (tcp://localhost:3000) [hats]  
 3342 s002  S+     0:00.18 puma: cluster worker 0: 3335 [hats]  
```

This seems independent from Rails though. If I start the server via `puma -b tcp://localhost:3000` and then `touch tmp/restart.txt` it yields the same issue.

If I start it with `puma -b tcp://127.0.0.1:3000`, it has no problem.

My guess is something to do with IPv6? I tried commenting out `::1 localhost` in `/etc/hosts` with no luck. I didn't reboot or do anything to clear my DNS though.

This is far as my journey goes. 😄 I hope this info helps.
